src/tests/Microsoft.PowerFx.Interpreter.Tests/RecalcEngineTests.cs

```diff
@@ -410,307 +409,128 @@ public void FormulaCantRedefine()
-        {
-            var config = new PowerFxConfig();
-            var recalcEngine = new RecalcEngine(config);
-            Assert.False(recalcEngine.DefineFunctions("foo():Blank = foo();").Errors.Any());
-            var result = recalcEngine.Eval("foo()");
-            Assert.IsType<ErrorValue>(result);
-        }
-
-        [Fact]
-        public void DefHailstoneSequence()
+        [Theory]
+        [InlineData(
+            "func1(x:Number/*comment*/): Number = x * 10;\nfunc2(x:Number): Number = y1 * 10;",
+            null,
+            true,
+            null)]
+        [InlineData(
+            "foo(x:Number):Number = If(x=0,foo(1),If(x=1,foo(2),If(x=2,Float(2))));",
+            "foo(Float(0))",
+            false,
+            2.0)]
+        [InlineData(
+            "foo(x:Decimal):Decimal = If(x=0,foo(1),If(x=1,foo(2),If(x=2,2)));",
+            "foo(0)",
+            false,
+            2,
+            true)]
+        [InlineData(
+            "foo():Blank = foo();",
+            "foo()",
+            true,
+            null)]
+
+        // Recursive calls are not allowed
+        [InlineData(
+            "hailstone(x:Number):Number = If(Not(x = 1), If(Mod(x, 2)=0, hailstone(x/2), hailstone(3*x+1)), x);",
+            "hailstone(Float(192))",
+            true,
+            null)]
+        [InlineData(
+            "hailstone(x:Decimal):Decimal = If(Not(x = 1), If(Mod(x, 2)=0, hailstone(x/2), hailstone(3*x+1)), x);",
+            "hailstone(192)",
+            true,
+            null)]
+        [InlineData(
+            "odd(number:Number):Boolean = If(number = 0, false, even(Abs(number)-1)); even(number:Number):Boolean = If(number = 0, true, odd(Abs(number)-1));",
+            "odd(17)",
+            true,
+            null)]
+        [InlineData(
+            "odd(number:Number):Boolean = If(number = 0, false, even(Abs(number)-1)); even(number:Number):Boolean = If(number = 0, true, odd(Abs(number)-1));",
+            "even(17)",
+            true,
+            null)]
+        [InlineData(
+            "odd(number:Decimal):Boolean = If(number = 0, false, even(If(number<0,-number,number)-1)); even(number:Decimal):Boolean = If(number = 0, true, odd(If(number<0,-number,number)-1));",
+            "odd(17)",
+            true,
+            null,
+            true)]
+        [InlineData(
+            "odd(number:Decimal):Boolean = If(number = 0, false, even(If(number<0,-number,number)-1)); even(number:Decimal):Boolean = If(number = 0, true, odd(If(number<0,-number,number)-1));",
+            "even(17)",
+            true,
+            null,
+            true)]
+
+        // Redefinition is not allowed
+        [InlineData(
+            "foo():Blank = foo(); foo():Number = x + 1;", 
+            null,
+            true,
+            null)]
+
+        // Syntax error
+        [InlineData(
+            "foo():Blank = x[",
+            null,
+            true,
+            null)]
+
+        // Incorrect parameters

```

Comment:
>// Incorrect parameters [](http://example.com/codeflow?start=8&length=23)

Can we please add some tests with coercion?

Reply:
I'll leave this at it is for now since we are moving this to txt files.